DEP: deprecate rollaxis

nschloe opened the pull request:
The changelog of 1.11 says about the then-new moveaxis:

> This function should be easier to use than the current rollaxis function as well as providing more functionality.

[I've recently stumbled upon rollaxis's weirdness myself](https://github.com/numpy/numpy/issues/9473), so perhaps it's time to deprecate it.

This PR replaces all occurrences of `rollaxis` by its corresponding `moveaxis` equivalent, and arguably the code becomes clearer. For example, instead of
```python
c = np.rollaxis(c, iaxis)
# do something
c = np.rollaxis(c, 0, iaxis + 1)
```
the new
```python
c = np.moveaxis(c, iaxis, 0)
# do something
c = np.moveaxis(c, 0, iaxis)
```
makes clear that `iaxis` moved to the first spot, then something happens, and then it's moved back.

nschloe committed:
replace rollaxis by moveaxis in numpy.polynomial

nschloe committed:
replace rollaxis by moveaxis everywhere except rollaxis and its tests

nschloe committed:
deprecate rollaxis()

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -2006,7 +2006,7 @@ def _multi_svd_norm(x, row_axis, col_axis, op):
     if row_axis > col_axis:
         row_axis -= 1
-    y = rollaxis(rollaxis(x, col_axis, x.ndim), row_axis, -1)
+    y = moveaxis(moveaxis(x, col_axis, -1), row_axis, -2)
```
This can be done with a single call to moveaxis

shoyer commented:
I'm +1 for moving internal uses of rollaxis to moveaxis.

I'm -1 on actually deprecating rollaxis. A louder warning in rollaxis against using it would be great (e.g., "We recommend using more moveaxis instead."). Take a look at the docs for `hstack`, for example: https://docs.scipy.org/doc/numpy/reference/generated/numpy.hstack.html

We don't actually remove functionality in NumPy unless absolutely necessary, and we don't deprecate anything unless we actually plan to remove it. `rollaxis` has a confusing API but keeping it with a warning is pretty harmless.

eric-wieser commented:
Don't forget the DEP: and MAINT: commit prefices. Also, this fails because you're not hiding deprecation warnings in rollaxis tests

eric-wieser commented:
@shoyer: to be clear, by warning you mean documentation-only, as opposed to an actual python warning.

eric-wieser commented:
Perhaps we should introduce a `OutdatedWarning` that's indicates a feature is there to stay, but has a better replacement?

charris commented:
I'm with @shoyer on this one.

nschloe committed:
DEP don't formally deprecate rollaxis, just add a hint to the documentation

shoyer commented:
> Perhaps we should introduce a OutdatedWarning that's indicates a feature is there to stay, but has a better replacement?

Sure, but only if we can silence the warning by default. Otherwise it's an even more annoying version of DeprecationWarning -- one that's entirely useless!

A red box like what you find in the Python docs for deprecated functions could be a nice approach:
https://docs.python.org/2/library/string.html#deprecated-string-functions

(Notice that none of these functions actually issue a warning when you use them.)

nschloe committed:
MAINT replace two moveaxis by one

nschloe commented:
Thanks for the comments, I've addressed the issues mentioned so far.

As for the deprecation, I would say there must be _some_ sort of it. This little text found in `hstack` docstring (and `rollaxis` after merging this PR) is hardly enough for me to notice it, so I'd vote for something more striking. If you don't want to annoy users with console messages, perhaps @shoyer's suggestiong of red boxes is the way to do. Seems pythonic anyways.

Should we tackle this now or on another PR?

shoyer commented:
> Should we tackle this now or on another PR?

Maybe another PR? There's a long list of semi-deprecated functions. Clarifying that could require some discussion.

nschloe commented:
I've added #9478 so we don't lose track of this.

eric-wieser commented on numpy/core/numeric.py:
```diff
@@ -1436,6 +1436,10 @@ def rollaxis(a, axis, start=0):
 
+    This function continues to be supported for backward compatibility, but you
+    should prefer np.moveaxis. The np.moveaxis function was added in NumPy
+    1.11.
```
referring to this as `` `moveaxis` `` will cause a link to be generated

nschloe replied:
Is it because of the missing tick or the additional `np.`?

eric-wieser replied:
Both, I think. There's nothing special about "np" - remember, the module's called "numpy" (but that's implied anyway in references)

nschloe replied:
Got it.

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -2006,7 +2006,7 @@ def _multi_svd_norm(x, row_axis, col_axis, op):
     if row_axis > col_axis:
         row_axis -= 1
-    y = rollaxis(rollaxis(x, col_axis, x.ndim), row_axis, -1)
+    y = moveaxis(x, [row_axis, col_axis], [-2, -1])
```
Might be inclined to pass tuples here, if only because the function used internally is called `normalize_index_tuple`

nschloe replied:
Also, the tests are failing here. travis says

> ValueError: repeated axis in `source` argument

so I'm guessing this occurs when `row_axis` and `col_axis` are equal. Perhaps we _do_ need two `moveaxis` here.

nschloe replied:
I'm a bit lost as to why `if row_axis > col_axis: row_axis -= 1` is done.

eric-wieser replied:
Looks like a crutch for the rollaxis behaviour to me

nschloe replied:
The commit that introduced this function is 4 years old, a2926149891027fff151d50933405a696384d47c, and the code hasn't been touched since.

eric-wieser replied:
And in fact, after removing it, there's an assertion for nonequality at the caller

eric-wieser replied:
So to conclude: just remove those two lines

nschloe replied:
Ah yes, I get it now: If `if row_axis > col_axis`, and `col_axis` is moved to the end of the shape first, one needs to `row_axis -= 1` to still get the right axis when moving the second one to the end.

Indeed, we don't need this anymore. Praised be `moveaxis`!

nschloe committed:
MAINT remove workaround for rollaxis

nschloe committed:
better docstring

eric-wieser reviewed (approved):
LGTM, but needs squashing - we can do that from GitHub's ui if needed

charris commented:
@nschloe  Could you add a comment to `doc/release/1.14.0-notes.rst`?

nschloe commented:
@charris  Under _deprecations_?

charris commented:
Oh, sorry. I see that no actual warning is issued, so no note is needed.

mhvk commented:
Late here, but hurray for this cleanup! (and for changing the main documentation to make it clearer that `rollaxis` is no longer something we recommend people to use)

charris commented:
@eric-wieser Want to finish this up?

eric-wieser commented:
@charris: Thought I'd give time for @nschloe to do the squash themself

eric-wieser commented:
Thanks @nschloe!

eric-wieser commented:
Interestingly this seems to cause a significant [performance decrease](https://pv.github.io/numpy-bench/#bench_function_base.Median.time_even_small?commits=91639937-10ed90c3&commits=320667d4-ddc53885&commits=aab7993a&commits=e332ba43&commits=029863ea). I don't think we should reverse this, but maybe `move_axis` should be made faster somehow.
